doc/sphinx/dev_git_intro.rst

```diff
@@ -52,46 +52,211 @@ Clone a local repository onto your machine
+4. Test the branch with the MDTF framework software
 
-   - If you encounter problems during practice, you can first try looking for *plain English* instructions to unmess the situation at `Dangit, Git?! <https://dangitgit.com/>`__.
+5. Check out your feature branch, then merge the test branch into the feature branch
+::
+  git checkout feature/<my_feature_name>
+  git merge <test_branch_name>
 
-- Write your code! A useful command is ``git status`` to remind you what branch you're on and changes you've made (but have not committed yet). ``git branch -a`` lists all branches with ``*`` indicating the branch you're on.
+6. Push the updates to your remote branch
+::
+  git push origin feature/<my_feature_name>
 
-.. (TODO: tests ...)
-.. (TODO: adding files...)
-.. (- Commit changes with ``git commit -m <your commit message>``.) Somehow -m never works for YH.
-.. Good commit messages are key to making the project's history useful. To make this easier, instead of using the ``-m`` flag, you can configure git to launch your text editor of choice with ``git config --global core.editor "<command string to launch your editor>"``.
-.. - To provide further information, add a blank line after the summary and wrap text to 72 columns if your editor supports it (this makes things display nicer on some tools). Here's an `example <https://github.com/NOAA-GFDL/MDTF-diagnostics/commit/225b29f30872b60621a5f1c55a9f75bbcf192e0b>`__.
+7. Delete the test branch
+::
+  git branch -D <test_branch_name>
+
+Updating your feature branch by rebasing it onto the develop branch (preferred method)
+--------------------------------------------------------------------------------------
+Rebasing is procedure to integrate the changes from one branch into another branch. ``git rebase`` differs from ``git merge`` in that it reorders the commit history so that commits from the branch that is being updated are moved to the `tip` of the branch. This makes it easier to isolate changes in the feature branch, and usually results in fewer merge conflicts when the feature branch is merged into the develop branch.
+
+1. Update the local and remote develop branches on your fork as described steps 1--3  
+of the **Start Coding** section, then check out your feature branch
+::
+  git checkout feature/<my_feature_name>
+
+and launch an interactive rebase of your branch onto the develop branch.
+::
+  git rebase -i develop
+2. Your text editor will open in the terminal (Vim by default)
+and display your commit hashes with the oldest commit at the top
+::
+  pick 39n3b42 oldest commit
+  pick 320cnyn older commit
+  pick 20ac93c newest commit
+
+You may squash commits by replacing *pick* with *squash* for the commit(s) that are newer than the commit you want to combine with (i.e., the commits below the target commit).
```

Comment:
I don't understand why I would squash some or pick others. Advice here would be appreciated! The following section is great examples of what will happen, but I want the 'why'. Or best practices...

Reply:
@bitterbark You would squash commits that you want to combine--typically those that make related changes. For example, if you add a function in one commit (we'll call it abcdefg), then continue developing the function in the next commit (1234567), you might want to squash 1234567 and abcdefg:
```
pick abcdefg (newer)
squash 1234567 (older)
````
You can also squash all of your commits during a rebase onto the develop branch so that the log has a single message with your work after the most recent develop commit:
```
pick 
squash 
squash
squash
...
squash (newest commit)
```
This can involve more work to resolve merge conflicts if the develop branch has a lot of updates. The benefits of doing so result in a cleaner commit history, making it easier to isolate the changes you make from the ones we make. This seems trivial, but imagine (or remember, if you've already had this exact problem like I have) trying to debug an issue that arises while testing, say, a beta release, and your commits are mixed with the develop commits in the order they were made. It's way more difficult to debug.

Reply:
It makes good sense to use the latter procedure: squash all our commits into a single message. Definitely cleans it up from your end, and allows us to commit frequently with messages that are useful to us without worrying that they'll clutter the development history.

It will be more likely that developers end up using this strategy if it is the only pick/squash one presented in the documentation. (Although the other examples are interesting, they don't seem relevant to what you want us to do, right?)

Reply:
In general,  squashing all commits during a rebase is the preferred strategy, but I don't think it should be required in case users have a situation when doing so creates an overwhelming amount of merge conflicts, or commits can't be combined for some reason. I've had cases where merge commits (commits that mark merging changes from one branch into another) won't combine with the others on branch because there was an empty message or something,

Reply:
It's actually better to have lots of smaller commits.  [Here's one opinion.](https://medium.com/better-programming/why-you-should-write-small-git-commits-c9a042737aa6#:~:text=Advantages%20of%20Small%20Commits%20and%20Continuous%20Improvement&text=Revert%20a%20commit%20with%20ease,Write%20better%20commit%20messages.)

The idea is to be able to leverage `git revert` down the line.  Say you make four different changes to your code.  We'll call them A, B, C, and D.  You uncover many months from now that there's a mistake in C, but A B and D are fine.  You can simply call `git revert C` to remove that edit directly.  Otherwise you need to edit the entire file, make a new commit, etc.

Smaller commits also improve git's intrinsic merging strategies making merge conflicts less likely to occur.